hermes/tns.py

```diff
@@ -120,17 +123,86 @@ def convert_flux_units(hermes_units):
 
 
-def convert_hermes_message_to_tns(hermes_message, filenames_mapping):
-    """ Converts from hermes message format into TNS at report format """
-    # TODO: Add support for associated files
-    # TODO: Add support for classification or frb reports
+def convert_classification_hermes_message_to_tns(hermes_message, target_filenames_mapping, spectroscopy_filenames_mapping):
+    """ Converts from a hermes message format into a TNS classification report format """
+    report_payload = {}
+    tns_options = get_reverse_tns_values()
+    data = hermes_message.get('data', {})
+    spectroscopy_by_target = defaultdict(list)
+    for spectra in data.get('spectroscopy', []):
+        spectroscopy_by_target[spectra['target_name']].append(spectra)
+    targets_by_name = {target['name']: target for target in data.get('targets', [])}
+
+    for k, target in enumerate(targets_by_name.values()):
+        if target['name'] in spectroscopy_by_target:
+            # This means we have at least one spectroscopy datum for this target, so make a classification report from it
+            classification_report = {
+                'related_files': {},
+                'spectra': {'spectra-group': {}}
+            }
+            discovery_info = target.get('discovery_info', {})
+            groups = target.get('group_associations', [])
+            classification_report['name'] = target['name']
+            classification_report['classifier'] = hermes_message.get('authors')
+            classification_report['groupid'] = str(tns_options.get('groups', {}).get(discovery_info.get('reporting_group'), -1))
+            classification_report['class_proprietary_period_groups'] = [str(tns_options.get('groups', {}).get(group, -1)) for group in groups]
+            classification_report['remarks'] = target.get('comments', '')
+            if target.get('redshift'):
+                classification_report['redshift'] = target.get('redshift')
+
+            first_spectra = spectroscopy_by_target[target['name']][0]
+            # Set classification object_type from the first spectrum
+            classification_report['objtypeid'] = str(tns_options.get('object_types', {}).get(first_spectra.get('classification'), -1))
+            # Proprietary period of the classification uses the targets discovery info proprietary period but should be left to 0 usually
+            classification_report['class_proprietary_period'] = {
+                'class_proprietary_period_value': str(discovery_info.get('proprietary_period', 0)),
+                'class_proprietary_period_units': discovery_info.get('proprietary_period_units', 'Days').lower()
+            }
+            for i, spectra in enumerate(spectroscopy_by_target[target['name']]):
+                spectra_report = {
+                    'obsdate': parse_date(spectra.get('date_obs')).strftime('%Y-%m-%d %H:%M:%S'),
+                    'instrumentid': str(tns_options.get('instruments', {}).get(spectra.get('instrument'))),
```

Comment:
Is it worth functionizing this pattern? It seems like we have a lot of repeated code here.